File: app/src/test/java/com/kaciula/archiman/data/UserRepositoryImplTest.java
package com.kaciula.archiman.data;

import static org.hamcrest.core.Is.is;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.kaciula.archiman.domain.entity.User;
import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import java.util.ArrayList;
import java.util.List;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

public class UserRepositoryImplTest {

  @Mock UserDataSource usersRemoteDataSource;
  @Mock UserDataSource usersLocalDataSource;

  private UserRepositoryImpl usersRepository;

  @Before
  public void setupUseCase() {
    MockitoAnnotations.initMocks(this);

    usersRepository = new UserRepositoryImpl(usersRemoteDataSource, usersLocalDataSource);
  }

  @Test
  public void getUsers() {
    final List<User> users = new ArrayList<>();
    users.add(User.create(1, "Best programmer"));
    users.add(User.create(2, "Second best programmer"));
    when(usersLocalDataSource.getUsers()).thenReturn(Observable.<List<User>>empty());
    when(usersLocalDataSource.createOrUpdateUsers(users)).thenReturn(Completable.complete());
    when(usersRemoteDataSource.getUsers()).thenReturn(Observable.fromArray(users));

    usersRepository.getUsers()
        .subscribe(new Observer<List<User>>() {
                     @Override
                     public void onSubscribe(Disposable d) {
                       // do nothing
                     }

                     @Override
                     public void onNext(List<User> usersList) {
                       Assert.assertThat(usersList, is(users));
                     }

                     @Override
                     public void onError(Throwable e) {
                       Assert.fail("Failed to receive the users list");
                     }

                     @Override
                     public void onComplete() {
                       // do nothing
                     }
                   }
        );

    verify(usersRemoteDataSource).getUsers();
  }
}

File: app/src/test/java/com/kaciula/archiman/data/UserRepositoryImplTest.kt
package com.kaciula.archiman.data

import com.kaciula.archiman.domain.entity.User
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import org.hamcrest.core.Is.`is`
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations
import java.util.*

class UserRepositoryImplTest {

    @Mock lateinit var usersRemoteDataStore: UserDataStore
    @Mock lateinit var usersLocalDataStore: UserDataStore

    private lateinit var usersRepository: UserRepositoryImpl

    @Before
    fun setupUseCase() {
        MockitoAnnotations.initMocks(this)

        usersRepository = UserRepositoryImpl(usersRemoteDataStore, usersLocalDataStore)
    }

    @Test
    fun getUsers() {
        val users = ArrayList<User>()
        users.add(User(1, "Best programmer"))
        users.add(User(2, "Second best programmer"))
        `when`(usersLocalDataStore.getUsers()).thenReturn(Observable.empty<List<User>>())
        `when`(usersLocalDataStore.createOrUpdateUsers(users)).thenReturn(Completable.complete())
        `when`(usersRemoteDataStore.getUsers()).thenReturn(Observable.fromArray<List<User>>(users))

        usersRepository.getUsers()
                .subscribe(object : Observer<List<User>> {
                    override fun onSubscribe(d: Disposable) {
                        // do nothing
                    }

                    override fun onNext(usersList: List<User>) {
                        Assert.assertThat(usersList, `is`<List<User>>(users))
                    }

                    override fun onError(e: Throwable) {
                        Assert.fail("Failed to receive the users list")
                    }

                    override fun onComplete() {
                        // do nothing
                    }
                }
                )

        verify<UserDataStore>(usersRemoteDataStore).getUsers()
    }
}
